dowhy/gcm/_noise.py

```diff
@@ -20,3 +40,108 @@ def compute_noise_from_data(causal_model: InvertibleStructuralCausalModel, obser
                     observed_data[get_ordered_predecessors(causal_model.graph, node)].to_numpy()))
 
     return result
+
+
+def get_noise_dependent_function(causal_model: StructuralCausalModel,
+                                 target_node: Any,
+                                 approx_prediction_model: Optional[PredictionModel] = None,
+                                 num_training_samples: int = 20000) \
+        -> Tuple[Callable[[np.ndarray], np.ndarray], List[Any]]:
+    """Returns a function that represent the given target_node and that is only dependent on upstream noise nodes.
+    This is, Y = f(N_0, N_1, ..., N_n), where Y is the target node and N_i the noise node of an upstream node. Since
+    the order of the noise variables can be ambiguous, this method also returns a list with the expected order of the
+    noise variables indicated by the name of the corresponding node. For instance:
+    Lets say we have a target variable X4 which has only 2 other upstream variables X1 and X3, then this methods returns
+    a callable that represents X4 = f(N_1, N_3, N_4). This callable expects a numpy array as input. Since the order
+    of the columns in this array is unclear, the method also returns a list with the names of the corresponding columns,
+    e.g. ['X1', 'X3', 'X4']. Note that the noise of X4 will also be an input variable here.
+
+    If an approx_prediction_model is given, the model is used instead to train a model from scratch to represent f.
+    Typically, it would be much faster to evaluate this model than propagating through the graph. The model is trained
```

Comment:
"Typically, it would be much faster to evaluate this model than propagating through the graph. "
Did not understand what you mean by evaluate versus propagating

Reply:
Lets say we have the causal graph ```X0 -> X1 -> X2 -> Y``` and are interested in sampling from ```Y``` after e.g. an intervention on an upstream node. Then we can either follow the topological order in the graph and execute the functional causal models one by one with the outputs from parent nodes or we learn one new model that represents Y with respect to its upstream noise terms. Here, this would be ```Y = f(X0, N1, N2, NY)``` . This model could be any kind of regression model (e.g. linear) and evaluating this model is then significantly faster for obtaining samples from ```Y``` than 'propagating' changes through the graph by evaluating the FCMs at each node.

A fast estimation here is particular important in many of our attribution algorithms where we nee to evaluate a lot of counterfactual outcomes.

Reply:
Got it, thanks!